### PR TITLE
SW-664 force to take pictures in camera settings

### PR DESCRIPTION
**Pull request by JFSolution**
- force to take a picture in camera settings
- fix mrb_state live update

### CODE REVIEW

File: octoprint_mrbeam/__init__.py [hunk 1]
```diff
@@ -2318,8 +2318,11 @@ def take_undistorted_picture(self, is_initial_calibration):
             "New undistorted image is requested. is_initial_calibration: %s",
             is_initial_calibration,
         )
-        self.lid_handler._photo_creator.is_initial_calibration = is_initial_calibration
-        self.lid_handler._startStopCamera("initial_calibration")
+        if is_initial_calibration:
+            self.lid_handler._photo_creator.is_initial_calibration = (
+                is_initial_calibration
+            )
+            self.lid_handler._startStopCamera("initial_calibration")
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/__init__.py]: Can we create an enum/variable with the event/string? then we can use it on the lid_handler as well as here. Otherwise you really need to know what the method is doing and it could lead to problems soon if we forget to change the string in one of the places...

File: octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js [hunk 2]
```diff
@@ -196,6 +196,15 @@ $(function () {
                 self.fromCurrentData = function (data) {
                     self._fromData(data);
                 };
+
+                self.onDataUpdaterPluginMessage = function (plugin, data) {
+                    if (plugin != "mrbeam") {
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js]: Could we make it a constant somewhere? I think we potentially use it in many places

File: octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js [hunk 3]
```diff
@@ -196,6 +196,15 @@ $(function () {
...
+                        return;
+                    }
+                    if ("mrb_state" in data) {
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js]: same here could we make it a constant? name could be state_key 🤷

File: octoprint_mrbeam/static/js/settings/camera_settings.js [hunk 4]
```diff
@@ -318,6 +320,23 @@ $(function () {
...
+                    self.forceTakeNewPicture()
+                }, 3000);
+            }else{
+                if (self.imageReloadTimer){clearInterval(self.imageReloadTimer)}
+            }
+        }
+
+        self.forceTakeNewPicture = function(){
+            if (self.readyToLaser.lid_fully_open() && !self.readyToLaser.state.interlocksClosed()){
+                OctoPrint.simpleApiCommand("mrbeam", "take_undistorted_picture", {})
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/settings/camera_settings.js]: Same here could we make a class or file with all this potential events? it would also help us visualise all the things we use from octoprint. We could start by creating the class and adding this one event on this PR and slowly add more or create a ticket to research other potential events

File: octoprint_mrbeam/__init__.py [hunk 5]
```diff
@@ -860,10 +861,12 @@ def on_ui_render(self, now, request, render_kwargs):
...
-                settings_model=SettingsService(self._logger, DocumentService(self._logger)).get_template_settings_model(
-                    self.get_model_id()),
-                burger_menu_model=BurgerMenuService(self._logger, DocumentService(self._logger)).get_burger_menu_model(
-                    self.get_model_id()),
+                settings_model=SettingsService(
+                    self._logger, DocumentService(self._logger)
+                ).get_template_settings_model(self.get_model_id()),
+                burger_menu_model=BurgerMenuService(
+                    self._logger, DocumentService(self._logger)
+                ).get_burger_menu_model(self.get_model_id()),
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/__init__.py]: This will give conflicts on my PR 😢

File: octoprint_mrbeam/static/js/app/view-models/mrbeam-constants.js [hunk 6]
```diff
@@ -0,0 +1,4 @@
+const MRBEAM = {
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/app/view-models/mrbeam-constants.js:1]: 👏

File: octoprint_mrbeam/static/js/app/view-models/mrbeam-simple-api-commands.js [hunk 7]
```diff
@@ -0,0 +1,5 @@
+const SimpleApiCommands = {
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/app/view-models/mrbeam-simple-api-commands.js:1]: 👏

File: octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js [hunk 8]
```diff
@@ -196,6 +196,15 @@ $(function () {
                 self.fromCurrentData = function (data) {
                     self._fromData(data);
                 };
+
+                self.onDataUpdaterPluginMessage = function (plugin, data) {
+                    if (plugin !== MRBEAM.PLUGIN_IDENTIFIER) {
+                        return;
+                    }
+                    if (MRBEAM.STATE_KEY in data) {
+                        self._fromData(data, "onDataUpdaterPluginMessage");
```
Comment (Yelko-mrbeam) [inline octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js:205]: Shouldn't this one also be part of the events? 🤔 I am not sure why is this needed since in the method doesn't seem to be used or? 😕

File: octoprint_mrbeam/__init__.py (hunk 5 above)
Comment (JFSolution) [reply octoprint_mrbeam/__init__.py]: Ohh I can revert, this was not intended to end up here 😅

File: octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js (hunk 8 above)
Comment (JFSolution) [reply octoprint_mrbeam/static/js/ready_to_laser_viewmodel.js:205]: I added "onDataUpdaterPluginMessage" so we know where we got the state from, I would say we can leave this as a string

**Review by Yelko-mrbeam — APPROVED** [review]
